### PR TITLE
docs: add info about tailcalls in bpf subprograms

### PR DESCRIPTION
**Pull request by mfijalko**
Here is an update to Cilium's documentation in order to reflect the current state of relationship between bpf2bpf calls and tail calls.
I tried to mimic as much as I could the style of provided image so that it follows all of the existing images.

Thanks!

### CODE REVIEW

File: Documentation/bpf.rst [hunk 1]
```diff
@@ -550,16 +550,44 @@ maximum number of nesting calls respectively allowed call frames is ``8``.
...
-the former adds additional stack frames and thus changes the expected layout
-for tail calls.
-
 BPF JIT compilers emit separate images for each function body and later fix
 up the function call addresses in the image in a final JIT pass. This has
 proven to require minimal changes to the JITs in that they can treat BPF to
 BPF calls as conventional BPF helper calls.
 
+Up to kernel 5.9, BPF tail calls and BPF subprograms excluded each other. BPF
+programs that utilized tail calls couldn't take the benefit of shrinked program
```
Comment (borkmann) [inline Documentation/bpf.rst]: nit: `benefit of shrinked program` -> `benefit of reducing program image size`

Comment (borkmann) [reply Documentation/bpf.rst]: Please also run `./Documentation/update-spelling_wordlist.sh func unwinded` so the latter two get added to the word list and don't cause doc CI failure. Thanks for extending the doc!

File: Documentation/bpf.rst [hunk 2]
```diff
@@ -550,16 +550,44 @@ maximum number of nesting calls respectively allowed call frames is ``8``.
...
+images and faster load times. Linux kernel 5.10 finally allows user to take the
+best of two worlds and adds the ability to combine the BPF subprograms with tail
+calls.
```
Comment (fristonio) [inline Documentation/bpf.rst]:
```suggestion
images and faster load times. Linux kernel 5.10 finally allows users to bring the best of two worlds and adds the ability to combine the BPF subprograms with tail calls.
```

File: Documentation/bpf.rst [hunk 3]
```diff
@@ -550,16 +550,44 @@ maximum number of nesting calls respectively allowed call frames is ``8``.
...
+
+This improvement comes with some restrictions, though. Mixing these two features
+could cause kernel stack overflow. To get an idea of what might happen, see the
```
Comment (fristonio) [inline Documentation/bpf.rst]:
```suggestion
This improvement comes with some restrictions, though. Mixing these two features
can cause a kernel stack overflow. To get an idea of what might happen, see the
```

File: Documentation/bpf.rst [hunk 4]
```diff
@@ -550,16 +550,44 @@ maximum number of nesting calls respectively allowed call frames is ``8``.
...
+
+.. image:: images/bpf_tailcall_subprograms.png
+    :align: center
+
+Tail call, before the actual jump to target program, will unwind only its
+current stack frame. As we can see on the example above, if a tail call takes
+place from within the sub function, the functions's (func1) stack frame will be
+present on stack when a program execution is at func2. Once the final (func3)
+function is terminated, all the previous stack frames will be unwinded and
+control will get back to the caller of BPF program.
```
Comment (fristonio) [inline Documentation/bpf.rst]:
```suggestion
Tail calls, before the actual jump to the target program, will unwind only its
current stack frame. As we can see in the example above, if a tail call occurs
from within the sub-function, the functions' (func1) stack frame will be
present on the stack when program execution is at func2. Once the final function(func3)
terminates, all the previous stack frames will be unwinded, and
control will get back to the BPF program caller.
```

File: Documentation/bpf.rst [hunk 5]
```diff
@@ -550,16 +550,44 @@ maximum number of nesting calls respectively allowed call frames is ``8``.
...
+256 bytes per each subprogram (note that if verifier detects the bpf2bpf call
+then main function is treated as a sub function as well). In total, with that
+restriction, BPF program's call chain can consume at most 8k of stack space.
+This limit comes from the 256 bytes per stack frame multiplied by 32 which is
+the tail call count limit. Without that, normally BPF programs can operate on
+512 byte stack size which would yield the 16k size in total for maximum count of
+tail calls and that solely would overflow the stack on some architectures.
+
+One more thing to mention is that this feature combination is currently
+supported only on x86-64 architecture.
```
Comment (fristonio) [inline Documentation/bpf.rst]: How about rewording this to something like?
```
The kernel introduced additional logic for detecting this feature combination.
There is a limit on the stack size throughout the whole call chain down to 256 bytes
per subprogram (note that if the verifier detects the bpf2bpf call, then the main
function is treated as a sub-function as well). In total, with this restriction,
the BPF program's call chain can consume at most 8KB of stack space.
This limit comes from the 256 bytes per stack frame multiplied by the tail call count limit(32).
Without this, the BPF programs will operate on 512-byte stack size, yielding the
16KB size in total for the maximum count of tail calls that would overflow the
stack on some architectures.
One more thing to mention is that this feature combination is currently supported only on the x86-64 architecture.
```

**Review by fristonio — CHANGES_REQUESTED** [review]
Thanks for the PR :rocket: 
Some minor improvements.

**Comment by joestringer** [comment]
@mfijalko thanks again for the submission :-) I've converted the PR to draft status for now. When you get a chance to address the review feedback, then please force-push the updated branch back here and mark the PR as "Ready for review" using the button at the bottom of the page.

**Comment by mfijalko** [comment]
> @mfijalko thanks again for the submission :-) I've converted the PR to draft status for now. When you get a chance to address the review feedback, then please force-push the updated branch back here and mark the PR as "Ready for review" using the button at the bottom of the page.

Ugh, sorry, that's embarrassing. Shortly after initial pull request submission I caught covid and totally forgot about that. Thanks for review, I'll try address comments in a day or two.

**Comment by joestringer** [comment]
@mfijalko No worries at all. I hope you're feeling better now :slightly_smiling_face:

**Comment by mfijalko** [comment]
Not sure if it's my fault that conformance test failed, but anyway I tried to address all of the review comments. Not sure where should I place versioning in github flow, commit message?

Anyway, hitting the 'ready for review' button.

**Review by borkmann — APPROVED** [review]

**Review by fristonio — APPROVED** [review]
LGTM :rocket: 
Thanks for sending out the PR.
